// @salesforce/source-deploy-retrieve/lib/src/client/metadataApiRetrieve.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MetadataApiRetrieve = exports.RetrieveResult = void 0;
/*
 * Copyright (c) 2021, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
const path = require("node:path");
const fs = require("graceful-fs");
const JSZip = require("jszip");
const ts_types_1 = require("@salesforce/ts-types");
const core_1 = require("@salesforce/core");
const kit_1 = require("@salesforce/kit");
const convert_1 = require("../convert");
const collections_1 = require("../collections");
const resolve_1 = require("../resolve");
const metadataTransfer_1 = require("./metadataTransfer");
const types_1 = require("./types");
const messages = new core_1.Messages('@salesforce/source-deploy-retrieve', 'sdr', new Map([["md_request_fail", "Metadata API request failed: %s"], ["error_convert_invalid_format", "Invalid conversion format '%s'"], ["error_could_not_infer_type", "%s: Could not infer a metadata type"], ["error_unexpected_child_type", "Unexpected child metadata [%s] found for parent type [%s]"], ["noParent", "Could not find parent type for %s (%s)"], ["error_expected_source_files", "%s: Expected source files for type '%s'"], ["error_failed_convert", "Component conversion failed: %s"], ["error_merge_metadata_target_unsupported", "Merge convert for metadata target format currently unsupported"], ["error_missing_adapter", "Missing adapter '%s' for metadata type '%s'"], ["error_missing_transformer", "Missing transformer '%s' for metadata type '%s'"], ["error_missing_type_definition", "Missing metadata type definition in registry for id '%s'."], ["error_missing_child_type_definition", "Type %s does not have a child type definition %s."], ["noChildTypes", "No child types found in registry for %s (reading %s at %s)"], ["error_no_metadata_xml_ignore", "Metadata xml file %s is forceignored but is required for %s."], ["noSourceIgnore", "%s metadata types require source files, but %s is forceignored."], ["noSourceIgnore.actions", "- Metadata types with content are composed of two files: a content file (ie MyApexClass.cls) and a -meta.xml file (i.e MyApexClass.cls-meta.xml). You must include both files in your .forceignore file. Or try appending \u201C\\*\u201D to your existing .forceignore entry.\n\nSee <https://developer.salesforce.com/docs/atlas.en-us.sfdx_dev.meta/sfdx_dev/sfdx_dev_exclude_source.htm> for examples"], ["error_path_not_found", "%s: File or folder not found"], ["noContentFound", "SourceComponent %s (metadata type = %s) is missing its content file."], ["noContentFound.actions", ["Ensure the content file exists in the expected location.", "If the content file is in your .forceignore file, ensure the meta-xml file is also ignored to completely exclude it."]], ["error_parsing_xml", "SourceComponent %s (metadata type = %s) does not have an associated metadata xml to parse"], ["error_expected_file_path", "%s: path is to a directory, expected a file"], ["error_expected_directory_path", "%s: path is to a file, expected a directory"], ["error_directory_not_found_or_not_directory", "%s: path is not a directory"], ["error_no_directory_stream", "%s doesn't support readable streams on directories."], ["error_no_source_to_deploy", "No source-backed components present in the package."], ["error_no_components_to_retrieve", "No components in the package to retrieve."], ["error_static_resource_expected_archive_type", "A StaticResource directory must have a content type of application/zip or application/jar - found %s for %s."], ["error_static_resource_missing_resource_file", "A StaticResource must have an associated .resource file, missing %s.resource-meta.xml"], ["error_no_job_id", "The %s operation is missing a job ID. Initialize an operation with an ID, or start a new job."], ["invalid_xml_parsing", "error parsing %s due to:\\n message: %s\\n line: %s\\n code: %s"], ["zipBufferError", "Zip buffer was not created during conversion"], ["undefinedComponentSet", "Unable to construct a componentSet. Check the logs for more information."], ["replacementsFileNotRead", "The file \"%s\" specified in the \"replacements\" property of sfdx-project.json could not be read."], ["unsupportedBundleType", "Unsupported Bundle Type: %s"], ["filePathGeneratorNoTypeSupport", "Type not supported for filepath generation: %s"], ["missingFolderType", "The registry has %s as is inFolder but it does not have a folderType"], ["tooManyFiles", "Multiple files found for path: %s."], ["cantGetName", "Unable to calculate fullName from path: %s (%s)"], ["missingMetaFileSuffix", "The metadata registry is configured incorrectly for %s. Expected a metaFileSuffix."], ["uniqueIdElementNotInRegistry", "No uniqueIdElement found in registry for %s (reading %s at %s)."], ["uniqueIdElementNotInChild", "The uniqueIdElement %s was not found the child (reading %s at %s)."], ["suggest_type_header", "A metadata type lookup for \"%s\" found the following close matches:"], ["suggest_type_did_you_mean", "-- Did you mean \".%s%s\" instead for the \"%s\" metadata type?"], ["suggest_type_more_suggestions", "Additional suggestions:\nConfirm the file name, extension, and directory names are correct. Validate against the registry at:\n<https://github.com/forcedotcom/source-deploy-retrieve/blob/main/src/registry/metadataRegistry.json>\n\nIf the type is not listed in the registry, check that it has Metadata API support via the Metadata Coverage Report:\n<https://developer.salesforce.com/docs/metadata-coverage>\n\nIf the type is available via Metadata API but not in the registry\n\n- Open an issue <https://github.com/forcedotcom/cli/issues>\n- Add the type via PR. Instructions: <https://github.com/forcedotcom/source-deploy-retrieve/blob/main/contributing/metadata.md>"]]));
const partialDeleteFileResponses = [];
class RetrieveResult {
    /**
     * @param response The metadata retrieve response from the server
     * @param components The ComponentSet of retrieved source components
     * @param localComponents The ComponentSet used to create the retrieve request
     */
    constructor(response, components, localComponents) {
        this.response = response;
        this.components = components;
        this.localComponents = new collections_1.ComponentSet(localComponents?.getSourceComponents());
    }
    getFileResponses() {
        if (this.response && this.fileResponses) {
            return this.fileResponses;
        }
        this.fileResponses = [];
        // construct failures
        if (this.response.messages) {
            const retrieveMessages = (0, kit_1.ensureArray)(this.response.messages);
            for (const message of retrieveMessages) {
                // match type name and fullname of problem component
                const matches = new RegExp(/.+'(.+)'.+'(.+)'/).exec(message.problem);
                if (matches) {
                    const [typeName, fullName] = matches.slice(1);
                    this.fileResponses.push({
                        fullName,
                        type: typeName,
                        state: types_1.ComponentStatus.Failed,
                        error: message.problem,
                        problemType: 'Error',
                    });
                }
                else {
                    this.fileResponses.push({
                        fullName: '',
                        type: '',
                        problemType: 'Error',
                        state: types_1.ComponentStatus.Failed,
                        error: message.problem,
                    });
                }
            }
        }
        // construct successes
        for (const retrievedComponent of this.components.getSourceComponents()) {
            const { fullName, type, xml } = retrievedComponent;
            const baseResponse = {
                fullName,
                type: type.name,
                state: this.localComponents.has(retrievedComponent) ? types_1.ComponentStatus.Changed : types_1.ComponentStatus.Created,
            };
            if (!type.children || Object.values(type.children.types).some((t) => t.unaddressableWithoutParent)) {
                for (const filePath of retrievedComponent.walkContent()) {
                    this.fileResponses.push(Object.assign({}, baseResponse, { filePath }));
                }
            }
            if (xml) {
                this.fileResponses.push(Object.assign({}, baseResponse, { filePath: xml }));
            }
        }
        // Add file responses for components that support partial delete (e.g., DigitalExperience)
        // where pieces of the component were deleted in the org, then retrieved.
        while (partialDeleteFileResponses.length) {
            const fromPartialDelete = partialDeleteFileResponses.pop();
            if (fromPartialDelete) {
                this.fileResponses.push(fromPartialDelete);
            }
        }
        return this.fileResponses;
    }
}
exports.RetrieveResult = RetrieveResult;
class MetadataApiRetrieve extends metadataTransfer_1.MetadataTransfer {
    constructor(options) {
        super(options);
        this.options = Object.assign({}, MetadataApiRetrieve.DEFAULT_OPTIONS, options);
    }
    /**
     * Check the status of the retrieve operation.
     *
     * @returns Status of the retrieve
     */
    async checkStatus() {
        if (!this.id) {
            throw new core_1.SfError(messages.getMessage('error_no_job_id', ['retrieve']), 'MissingJobIdError');
        }
        const coerceBoolean = (field) => {
            if ((0, ts_types_1.isString)(field)) {
                return field.toLowerCase() === 'true';
            }
            return (0, ts_types_1.asBoolean)(field, false);
        };
        const connection = await this.getConnection();
        // Cast RetrieveResult returned by jsForce to MetadataApiRetrieveStatus
        const status = (await connection.metadata.checkRetrieveStatus(this.id));
        status.fileProperties = (0, kit_1.ensureArray)(status.fileProperties);
        status.success = coerceBoolean(status.success);
        status.done = coerceBoolean(status.done);
        return status;
    }
    /**
     * Cancel the retrieve operation.
     *
     * Canceling a retrieve occurs immediately and requires no additional status
     * checks to the org, unlike {@link MetadataApiDeploy.cancel}.
     */
    // eslint-disable-next-line @typescript-eslint/require-await
    async cancel() {
        this.canceled = true;
    }
    async post(result) {
        let components;
        const isMdapiRetrieve = this.options.format === 'metadata';
        if (result.status === types_1.RequestStatus.Succeeded) {
            const zipFileContents = Buffer.from(result.zipFile, 'base64');
            if (isMdapiRetrieve) {
                const name = this.options.zipFileName ?? 'unpackaged.zip';
                const zipFilePath = path.join(this.options.output, name);
                fs.writeFileSync(zipFilePath, zipFileContents);
                if (this.options.unzip) {
                    const zip = await JSZip.loadAsync(zipFileContents, { base64: true, createFolders: true });
                    const extractPath = path.join(this.options.output, path.parse(name).name);
                    fs.mkdirSync(extractPath, { recursive: true });
                    for (const filePath of Object.keys(zip.files)) {
                        const zipObj = zip.file(filePath);
                        if (!zipObj || zipObj?.dir) {
                            fs.mkdirSync(path.join(extractPath, filePath), { recursive: true });
                        }
                        else {
                            // eslint-disable-next-line no-await-in-loop
                            const content = await zipObj?.async('nodebuffer');
                            if (content) {
                                fs.writeFileSync(path.join(extractPath, filePath), content);
                            }
                        }
                    }
                }
            }
            else {
                components = await this.extract(zipFileContents);
            }
        }
        components ??= new collections_1.ComponentSet(undefined, this.options.registry);
        const retrieveResult = new RetrieveResult(result, components, this.components);
        if (!isMdapiRetrieve && !this.options.suppressEvents) {
            // This should only be done when retrieving source format since retrieving
            // mdapi format has no conversion or events/hooks
            await this.maybeSaveTempDirectory('source', components);
            await core_1.Lifecycle.getInstance().emit('scopedPostRetrieve', {
                retrieveResult,
                orgId: this.orgId,
            });
        }
        return retrieveResult;
    }
    async pre() {
        const packageNames = this.getPackageNames();
        if (this.components?.size === 0 && !packageNames?.length) {
            throw new core_1.SfError(messages.getMessage('error_no_components_to_retrieve'), 'MetadataApiRetrieveError');
        }
        const connection = await this.getConnection();
        const apiVersion = connection.getApiVersion();
        this.orgId = connection.getAuthInfoFields().orgId;
        if (this.components) {
            this.components.apiVersion ??= apiVersion;
            this.components.sourceApiVersion ??= apiVersion;
        }
        // only do event hooks if source, (NOT a metadata format) retrieve
        if (this.options.components && !this.options.suppressEvents) {
            await core_1.Lifecycle.getInstance().emit('scopedPreRetrieve', {
                componentSet: this.options.components,
                orgId: this.orgId,
            });
        }
        const manifestData = (await this.components?.getObject())?.Package;
        const requestBody = {
            // This apiVersion is only used when the version in the package.xml (manifestData) is not defined.
            // see docs here: https://developer.salesforce.com/docs/atlas.en-us.api_meta.meta/api_meta/meta_retrieve_request.htm
            apiVersion: this.components?.sourceApiVersion ?? (await connection.retrieveMaxApiVersion()),
            ...(manifestData ? { unpackaged: manifestData } : {}),
        };
        // if we're retrieving with packageNames add it
        // otherwise don't - it causes errors if undefined or an empty array
        if (packageNames?.length) {
            requestBody.packageNames = packageNames;
            // delete unpackaged when no components and metadata format to prevent
            // sending an empty unpackaged manifest.
            if (this.options.format === 'metadata' && this.components?.size === 0) {
                delete requestBody.unpackaged;
            }
        }
        if (this.options.singlePackage) {
            requestBody.singlePackage = this.options.singlePackage;
        }
        // Debug output for API version used for retrieve
        const manifestVersion = manifestData?.version;
        if (manifestVersion) {
            this.logger.debug(`Retrieving source in v${manifestVersion} shape using SOAP v${apiVersion}`);
            await core_1.Lifecycle.getInstance().emit('apiVersionRetrieve', { manifestVersion, apiVersion });
        }
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore required callback
        return connection.metadata.retrieve(requestBody);
    }
    getPackageNames() {
        return this.getPackageOptions()?.map((pkg) => pkg.name);
    }
    getPackageOptions() {
        const { packageOptions } = this.options;
        return (packageOptions ?? []).map((po) => (0, ts_types_1.isString)(po) ? { name: po, outputDir: po } : { name: po.name, outputDir: po.outputDir ?? po.name });
    }
    async extract(zip) {
        let components = [];
        const { merge, output, registry } = this.options;
        const converter = new convert_1.MetadataConverter(registry);
        const tree = await resolve_1.ZipTreeContainer.create(zip);
        const packages = [{ zipTreeLocation: 'unpackaged', outputDir: output }].concat(this.getPackageOptions().map(({ name, outputDir }) => ({ zipTreeLocation: name, outputDir })));
        for (const pkg of packages) {
            const outputConfig = merge
                ? {
                    type: 'merge',
                    mergeWith: this.components?.getSourceComponents() ?? [],
                    defaultDirectory: pkg.outputDir,
                    forceIgnoredPaths: this.components?.forceIgnoredPaths ?? new Set(),
                }
                : {
                    type: 'directory',
                    outputDirectory: pkg.outputDir,
                };
            const zipComponents = collections_1.ComponentSet.fromSource({
                fsPaths: [pkg.zipTreeLocation],
                registry,
                tree,
            })
                .getSourceComponents()
                .toArray();
            if (merge) {
                this.handlePartialDeleteMerges(zipComponents, tree);
            }
            // this is intentional sequential
            // eslint-disable-next-line no-await-in-loop
            const convertResult = await converter.convert(zipComponents, 'source', outputConfig);
            components = components.concat(convertResult?.converted ?? []);
        }
        return new collections_1.ComponentSet(components, registry);
    }
    // Some bundle-like components can be partially deleted in the org, then retrieved. When this
    // happens, the deleted files need to be deleted on the file system and added to the FileResponses
    // that are returned by `RetrieveResult.getFileResponses()` for accuracy. The component types that
    // support this behavior are defined in the metadata registry with `"supportsPartialDelete": true`.
    // However, not all types can be partially deleted in the org. Currently this only applies to
    // DigitalExperienceBundle and ExperienceBundle.
    handlePartialDeleteMerges(retrievedComponents, tree) {
        const partialDeleteComponents = new Map();
        const mergeWithComponents = this.components?.getSourceComponents().toArray() ?? [];
        // Find all merge (local) components that support partial delete.
        mergeWithComponents.forEach((comp) => {
            if (comp.type.supportsPartialDelete && comp.content && fs.statSync(comp.content).isDirectory()) {
                const contentList = fs.readdirSync(comp.content);
                partialDeleteComponents.set(comp.fullName, { contentPath: comp.content, contentList });
            }
        });
        // If no partial delete components were in the mergeWith ComponentSet, no need to continue.
        if (partialDeleteComponents.size === 0) {
            return;
        }
        // Compare the contents of the retrieved components that support partial delete with the
        // matching merge components. If the merge components have files that the retrieved components
        // don't, delete the merge component and add all locally deleted files to the partial delete list
        // so that they are added to the `FileResponses` as deletes.
        retrievedComponents.forEach((comp) => {
            if (comp.type.supportsPartialDelete && partialDeleteComponents.has(comp.fullName)) {
                const localComp = partialDeleteComponents.get(comp.fullName);
                if (localComp?.contentPath && comp.content && tree.isDirectory(comp.content)) {
                    const remoteContentList = tree.readDirectory(comp.content);
                    const isForceIgnored = (filePath) => {
                        const ignored = comp.getForceIgnore().denies(filePath);
                        if (ignored) {
                            this.logger.debug(`Local component has ${filePath} while remote does not, but it is forceignored so ignoring.`);
                        }
                        return ignored;
                    };
                    localComp.contentList.forEach((fileName) => {
                        if (!remoteContentList.includes(fileName)) {
                            // If fileName is forceignored it is not counted as a diff. If fileName is a directory
                            // we have to read the contents to check forceignore status or we might get a false
                            // negative with `denies()` due to how the ignore library works.
                            const fileNameFullPath = path.join(localComp.contentPath, fileName);
                            if (fs.statSync(fileNameFullPath).isDirectory()) {
                                const nestedFiles = fs.readdirSync(fileNameFullPath);
                                if (nestedFiles.some((f) => isForceIgnored(path.join(fileNameFullPath, f)))) {
                                    return;
                                }
                            }
                            else if (isForceIgnored(fileNameFullPath)) {
                                return;
                            }
                            this.logger.debug(`Local component (${comp.fullName}) contains ${fileName} while remote component does not. This file is being removed.`);
                            const filePath = path.join(localComp.contentPath, fileName);
                            partialDeleteFileResponses.push({
                                fullName: comp.fullName,
                                type: comp.type.name,
                                state: types_1.ComponentStatus.Deleted,
                                filePath,
                            });
                            fs.rmSync(filePath, { recursive: true, force: true });
                        }
                    });
                }
            }
        });
    }
}
exports.MetadataApiRetrieve = MetadataApiRetrieve;
MetadataApiRetrieve.DEFAULT_OPTIONS = { merge: false };
